### PR TITLE
Introduce some debugging functions to dump arrays to disk

### PR DESCRIPTION
**Pull request by mabruzzo**
This is a small PR with some changes to introduce some basic functionality to help me finish resolving PR #260 (I hope to finish resolving that Issue this coming week).

Specifically, I added the functions: `disk_utils::dump_arrays_to_hdf5` and `disk_utils::dump_array_to_hdf5` to let us write arbitrary`CelloArray`s to disk at any time (i.e. in the middle of a method's execution). This is something I've we wished that we had multiple times in the past when I debugged other issues.

### CODE REVIEW

File: src/Cello/disk_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace disk_utils{
+
+
+namespace detail{ // helper functions not meant to be used elsewhere
+
+/// helper function that actually writes ``arr`` to an open HDF5 file
+///
+/// This assumes that ``file`` has been opened and that the appropriate group
+/// has been created (and is openned)
```
Comment (gregbryan) [inline src/Cello/disk_utils.hpp:21]:
```suggestion
/// has been created (and has already been opened)
```

**Review by gregbryan — APPROVED** [review]
Looks good to me!

**Comment by mabruzzo** [comment]
At some point in the last few months, I became aware that the `typeid` operation employs RTTI (runtime type inference) - which can apparently introduce runtime overhead. Thus, I decided to refactor ``cello::get_type_enum``. The function now operates at compile-time (by default it will provide a useful compile-time error if you try to give a type without a type_enum counterpart).

It was definitely possible to implement this in C++11. But, I realized that the implementation got a lot more concise (and easier to understand) in C++17. So I decided to do that instead. Consequently, this shouldn't get merged until after #325

**Review by bwoshea — APPROVED** [review]
Looks good to me!

**Pull request merged**